bigquery/google/cloud/bigquery/job.py

```diff
@@ -973,62 +978,126 @@ def from_api_repr(cls, resource, client):
         return job
 
 
-class _ExtractConfiguration(object):
-    """User-settable configuration options for extract jobs.
+class ExtractJobConfig(object):
+    """Configuration options for extract jobs.
 
-    Values which are ``None`` -> server defaults.
+    All properties in this class are optional. Values which are ``None`` ->
+    server defaults.
     """
-    _compression = None
-    _destination_format = None
-    _field_delimiter = None
-    _print_header = None
+
+    def __init__(self):
+        self._properties = {}
+
+    compression = Compression('compression', 'compression')
+    """See
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.extract.compression
+    """
+
+    destination_format = DestinationFormat(
+        'destination_format', 'destinationFormat')
+    """See
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.extract.destinationFormat
+    """
+
+    field_delimiter = _TypedApiResourceProperty(
+        'field_delimiter', 'fieldDelimiter', six.string_types)
+    """See
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.extract.fieldDelimiter
+    """
+
+    print_header = _TypedApiResourceProperty(
+        'print_header', 'printHeader', bool)
+    """See
+    https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration.extract.printHeader
+    """
+
+    def to_api_repr(self):
+        """Build an API representation of the extact job config.
+
+        :rtype: dict
+        :returns: A dictionary in the format used by the BigQuery API.
+        """
+        return copy.deepcopy(self._properties)
+
+    @classmethod
+    def from_api_repr(cls, resource):
+        """Factory: construct a job configuration given its API representation
+
+        :type resource: dict
+        :param resource:
+            An extract job configuration in the same representation as is
+            returned from the API.
+
+        :rtype: :class:`google.cloud.bigquery.job.ExtractJobConfig`
+        :returns: Configuration parsed from ``resource``.
+        """
+        config = cls()
+        config._properties = copy.deepcopy(resource)
+        return config
 
 
 class ExtractJob(_AsyncJob):
     """Asynchronous job: extract data from a table into Cloud Storage.
 
     :type job_id: str
-    :param job_id: the job's ID, within the project belonging to ``client``.
+    :param job_id: the job's ID
```

Comment:
(Really a comment on the whole class.) 

Users shouldn't be creating this directly; they should always call Client.extract_table, right? Then why do we expose it? I believe @jonparrott said somewhere recently that we shouldn't return instances of private classes to users, but isn't that the right thing here?

Reply:
If this was Java, I'd say we could have a public class but make the constructor private. Not sure how to express that in Python.